File: src/location.js
// extracted so it can be tested without requiring the worker runtime
export function getCodeSandboxLocation(path) {
  const prefix = 'https://codesandbox.io/s/github/'
  const parts = path.substr(1).split('/')

  if (parts.length === 0) return null
  if (parts.length === 1) return null
  if (parts.length === 2) return prefix + parts.join('/')
  if (parts.length === 3) return null

  if (parts[2] === 'tree') return prefix + parts.join('/')

  if (parts[2] === 'blob')
    return (
      prefix +
      [parts[0], parts[1], 'tree', parts[3]].join('/') +
      '?file=/' +
      parts.slice(4).join('/')
    )

  return null
}

File: internal/githubbox/location.go
// Package githubbox implements the redirects served by githubbox.com.
package githubbox

import "strings"

const prefix = "https://codesandbox.io/s/github/"

// getCodeSandboxLocation returns the CodeSandbox URL for a GitHub path, or "" when the path
// does not point at something CodeSandbox can open.
func getCodeSandboxLocation(path string) string {
	var parts []string
	for _, part := range strings.Split(strings.TrimPrefix(path, "/"), "/") {
		if part != "" {
			parts = append(parts, part)
		}
	}

	switch len(parts) {
	case 0, 1, 3:
		return ""
	case 2:
		return prefix + strings.Join(parts, "/")
	}

	if parts[2] == "tree" {
		return prefix + strings.Join(parts, "/")
	}

	if parts[2] == "blob" {
		return prefix +
			strings.Join([]string{parts[0], parts[1], "tree", parts[3]}, "/") +
			"?file=/" + strings.Join(parts[4:], "/")
	}

	return ""
}
